### ISSUE TITLE
Question: How does a web app know if it should create it's own desktop agent or wait for one to be injected from a container?

### ISSUE THREAD

**Issue by Roaders** [finos/FDC3#1243]
### Question Area
 - [ ] App Directory
 - [x] API
 - [ ] Context Data
 - [ ] Intents
 - [ ] Use Cases
 - [ ] Other

### Question

If I create a web app that can be either run by itself in a browser or can be ran within a container that injects an agent how does it know if it should create the agent or not?

The spec says that the agent should be set to `window.fdc3` and when it has been set an `fdc3Ready` event should be fired. How long should we wait for this event before creating an agent if we're not running in a container? Do we assume that when running in a container the agent is set and available before any code within the webpage can run? If this is the case what is the reason for the event?

It would be an annoyance if we need to wait a certain amount of time before creating an agent. Any code that interacts with the agent would need to first wait for the agent to resolved or created.

**Comment by kriswest** [comment]
In the FDC3 for Web Browsers proposal we are proposing a timeout, as one is not yet defined in the FDC3 standard. However, such things are difficult as timing can of course be affected by activity such as a market event, or reloading a layout containing many apps that are contending for CPU resources. 

> The spec says that the agent should be set to window.fdc3 and when it has been set an fdc3Ready event should be fired. How long should we wait for this event before creating an agent if we're not running in a container? Do we assume that when running in a container the agent is set and available before any code within the webpage can run? If this is the case what is the reason for the event?

Unfortunately, thats not currently a safe assumption. While it is possible in a container to get that reference created before other code runs (as one ofmy firm's products does), not every implemnetation does, nor are they required to by the Standard.

> It would be an annoyance if we need to wait a certain amount of time before creating an agent. Any code that interacts with the agent would need to first wait for the agent to resolved or created.

The proposed `getAgent()` function that would handle retrieving a Desktop Agent is expected to handle acquiring Desktop Agent's from a containers and 'parent' Desktop Agents in web browsers (i.e. DAs that created the page or frame within which the app is running and can be contacted at `window.opener`, `window.parent` or `window.parent.opener`) in parallel, returning the first it finds. If neither process identifies a Desktop Agent within the defined timeout then the `getAgent()` function will call any 'failover' function provided by the app which it can then use to create an agent or otherwise load an adaptor to one - returning either a Desktop Agent implementation or a window reference to contact.

This is not available today (beyond a POC that shows its a viable solution) as the proposal and associated PR need to be completed, then adopted by the FDC3 community. We are actively recruiting participants to help with refactoring/rewriting the POC into the final software solution to be added to the FDC3 NPM module as the group working to deliver the proposal is very small - although interest within the community appears to be high!

If you need to deliver something today, I would suggest using a 1 second timeout. Thats longer than we propose in FDC3 for Web browsers, but then there has never been a requirement set for Desktop Agent implementations to get this done faster. However, there has been the need (defined in the standard) for Desktop Agents to wait for apps to add their intent listeners before attempting to deliver an intent. There has been an attempt to tighten up the timeouts necessary on intent registration for 2.2, but after strong pushback from a Desktop Agent vendor, it was reduced to a recommendation (SHOULD) that an app add its intent listeners within 15 seconds of launch and that requirement that DAs MUST wait at least 15 seconds for an app to do so. I concede 15 seconds is very long time in web terms, but this also applies to native apps.

**Comment by Roaders** [comment]
Thanks for the detailed reply.
The spec states:

> Implementors MAY additionally make the API available through modules, imports, or other means.

We plan to do this. In fact we plan to make the agent available through an IOC container and inject it into interested components via the constructor. Unless we wait a second before constructing any of the app then I can't see how we would be able to get IOC to work with an agent.
15 seconds is a very long time but 1 second is still a long time to wait for your app to appear on the screen. 🙂

**Comment by kriswest** [comment]
Could you define IOC container for me?

Yes, depending on your context 15 seconds is along time - but the standard has to cater for every type of app so that minimum has to cater for a native app - and potentially a login screen (which others think 15 seconds is far too short for - however there are other ways to solve for that case, such as constructing but queuing handling of contexts and intents).

I suggest a second in the absence of a shorter defined timeout - as I mentioned its possible to get it doesn't before content loads and executes but we have no way of knowing if every product does that as its not required. It doesn't necessarily have o mean that you don't load or show the app, only that it needs to wait before handling interop - or better to queue calls. All FDC3 API calls are async so you can write code to all you to call functions immediately if you need to and only have them apply/resolve when the DA connection is ready

**Comment by kriswest** [comment]
P.S. Do remember that 'making the API available through modules, imports, or other means' is not enough for conformance to the standard - i.e. if won't work with 3rd party applications without specific changes to support your container if you also need to run those.

**Comment by Roaders** [comment]
> Could you define IOC container for me?

Sorry, Inversion of control. We want our agent to be injected into  the constructor of our classes so it is immediately available.
I suppose we could inject an agent proxy that waits for the real agent to be available as you suggest. If requests are made before window.fdc3 is available the call is queued until window.fdc3 has been set by a container or created.

**Comment by kriswest** [comment]
Heres the shared doc with current draft: https://tick42-my.sharepoint.com/:w:/g/personal/finsemble_datastore_interop_io/EZ0dfTCdRlJCnIF3C_1Oit0BF3fsXyvlMbisXp722DC9Kg?e=H2y7fn

It is at present a proposal, which I'm working on a specifications PR for (to go into the FDC3 Standard docs, as part of 2.2 if adopted), while @robmoffat and recently @julianna-ciq have been working on a POC of (which proves the techniques but doesn't yet match the letter of the proposal). There are also a set of schemas for all the messages that will need sending back and forth.

One point on terminology: The FDC3 Standard refers to th e Desktop Agent as a single entity managing multiple apps, with the API being an interface to it. FDC3 for the Web preserves that model by having you contact a 'parent' desktop agent (although it can also be a proxy to it of some sort). Presumably, you have the same model (a single agent multiple apps) - or have you gone for some more distributed with multiple cooperating agents?

> I suppose we could inject an agent proxy that waits for the real agent to be available as you suggest. If requests are made before window.fdc3 is available the call is queued until window.fdc3 has been set by a container or created.

Thats what I would recommend, although if FDC3 for Web Browsers is adopted you can swap that out later for just using the proposed `getAgent()` fn. Also ideally container DAs will move to doing something similar internally if they have setup time (queue requests until ready to deal with them), allowing them to inject the global ASAP - thats what we did and found it quite effective.

**Comment by Roaders** [comment]
We are planning on having one agent per browser window (by which I mean tabs, windows or Iframes). Probably we will have 1 app per window as well but we're not requiring that. They will communicate via window.postMessage and / or the channel api.

**Comment by kriswest** [comment]
👍 I think I've covered this in the other thread (#1242) - I would then think of your Desktop Agent as a distributed entity. I.e. its one agent, but it has parts running in multiple places. You may find it helpful in the long run to differentiate the API and that entity by referring to the Desktop Agent API and Desktop Agent or some other similar distinction as that will better match what you find in the FDC3 docs. FDC3 currently standardizes the API interface and expected behavior, but no implementation details. FDC3 for Web Browsers takes things further by adding a protocol (and handler function) for acquiring the different agent types, chosen communication mechanism (postMessage then channel messaging) and comms protocol (the actual messages that need sending).

The nature of the current FDC3 API for Web applications makes it hard to have more than one app per window as the API endpoint is a window global (window.fdc3) and it needs to implicitly know the identity of the application. There is a proposal in the doc I shared with you for FDC3 for the Web to resolve that by giving you a function to retrieve additional copies with different identities (`getSubAgent()` I think) and as I explained earlier, putting apps in iframes gives you something similar (as each iframe has its own DOM / scope). If thats something you will need, its worth raising an issue or adding a +1 - I'm not sure there  is a separate issue open for multiple apps in one window!

Finally, if you are implementing your own messaging please note that you could make use of the comms protocol as its largely complete already. The schemas can be found at: https://github.com/finos/FDC3/tree/fdc3-for-web/schemas/api, while some typescript types for them is generated at: https://github.com/finos/FDC3/blob/fdc3-for-web/src/api/BrowserTypes.ts
Theres a lot of documentation to write, but those messages are nominally complete and ready to use (and we'd love to see someone using them and identifying anything we forgot!)

**Comment by Roaders** [comment]
ok, thanks for that. I had started to think about and create the types for the messages but those should save me some time. I'll certainly take a look and try to use them.

**Comment by kriswest** [comment]
Great! Hopefully, these will be available from the FDC3 module some day soon, until then feel free to copy them over form the branch.

Closing this issue as answered - feel free to re-open if more questions occur
